nova/objects/instance.py

```diff
@@ -1553,20 +1553,25 @@ def _get_counts(instances):
                 if missing_itypes:
                     # Not found in flavor db. might be deleted. get it from the
                     # saved info
-                    iquery = context.session.query(
+                    # We fist get a list of instance UUIDs - one per flavor -
+                    # that use this flavor. Then we do a query against the
+                    # instance_extra table with those uuids. This is faster
+                    # than a subquery-JOIN, subquery with WHERE or a JOIN.
```

Comment:
Personally, I am not a fan of moving query to the application, but I can't reason against results :) I won't block this because of that.
Have you tried a CTE?

Reply:
I was not aware of CTEs, they make the query take 1.3s. by either joining the temporary table or by using IN with a subquery selecting from the temporary table.